cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,509 @@
+    return ptr_ >= other.ptr_;
+  }
+
+ private:
+  pointer ptr_ = NULLPTR;
+};
+
+template <typename T>
+class StaticVectorMixin {
+ protected:
+  // properly aligned uninitialized storage for N T's
+  using storage_type = typename std::aligned_storage<sizeof(T), alignof(T)>::type;
+
+  static T* ptr_at(storage_type* p, size_t i) {
+    return launder(reinterpret_cast<T*>(&p[i]));
+  }
+
+  static const T* ptr_at(const storage_type* p, size_t i) {
+    return launder(reinterpret_cast<const T*>(&p[i]));
+  }
+
+  static void move_storage(storage_type* src, storage_type* dest, size_t n) {
+    for (size_t i = 0; i < n; ++i) {
+      T* src_item = ptr_at(src, i);
+      T* dest_item = ptr_at(dest, i);
+      new (dest_item) T(std::move(*src_item));
+      src_item->~T();
+    }
+  }
+
+  static void destroy_storage(storage_type* p, size_t n) {
+    for (size_t i = 0; i < n; ++i) {
+      ptr_at(p, i)->~T();
+    }
+  }
+};
+
+template <typename T, size_t N, bool NonTrivialDestructor>
+class StaticVectorBaseStorage : public StaticVectorMixin<T> {
+ protected:
+  using typename StaticVectorMixin<T>::storage_type;
+
+  storage_type static_data_[N];
+  size_t size_ = 0;
+
+  void destroy() {}
+};
+
+template <typename T, size_t N>
+class StaticVectorBaseStorage<T, N, true> : public StaticVectorMixin<T> {
+ protected:
+  using typename StaticVectorMixin<T>::storage_type;
+
+  storage_type static_data_[N];
+  size_t size_ = 0;
+
+  ~StaticVectorBaseStorage() noexcept { destroy(); }
+
+  void destroy() noexcept { this->destroy_storage(static_data_, size_); }
+};
+
+template <typename T, size_t N, bool D = !std::is_trivially_destructible<T>::value>
+class StaticVectorBase : public StaticVectorBaseStorage<T, N, D> {
+ protected:
+  using Base = StaticVectorBaseStorage<T, N, D>;
+  using typename Base::storage_type;
+
+  using Base::size_;
+  using Base::static_data_;
+
+  StaticVectorBase() noexcept = default;
+
+  storage_type* data_ptr() { return static_data_; }
+
+  constexpr const storage_type* const_data_ptr() const { return static_data_; }
+
+  void bump_size(size_t addend) {
+    assert(size_ + addend <= N);
+    size_ += addend;
+  }
+
+  void reduce_size(size_t reduce_by) {
+    assert(reduce_by <= size_);
+    size_ -= reduce_by;
+  }
+
+  void move_from(StaticVectorBase&& other) noexcept {
+    size_ = other.size_;
+    this->move_storage(other.static_data_, static_data_, size_);
+    other.size_ = 0;
+  }
+
+ public:
+  constexpr size_t capacity() const { return N; }
+
+  constexpr size_t max_size() const { return N; }
+
+  void reserve(size_t n) {}
+
+  void clear() {
+    this->destroy_storage(static_data_, size_);
+    size_ = 0;
+  }
+};
+
+template <typename T, size_t N>
+class SmallVectorBase : public StaticVectorMixin<T> {
+ protected:
+  using typename StaticVectorMixin<T>::storage_type;
+
+  storage_type static_data_[N];
+  size_t size_ = 0;
+  storage_type* data_ = static_data_;
+  size_t dynamic_capacity_ = 0;
```

Comment:
This uses 24 bytes in addition to the static elements. I think it'd be worthwhile to compress this further. Here's a sketch which uses a minimum of 16 bytes (for example, `SmallVector<int32_t, 3>` would be only 16 bytes)

```c++
template <typename T, size_t N>
struct LittleEndianState {
  using Storage = StaticVectorMixin<T>;
  using storage_type = typename Storage::storage_type;

  struct DynamicState {
    // low bit will be zero for alignment
    T* data_;
    uint32_t size_, capacity_;
  };

  struct SmallState {
    // low bit will be one for discrimination
    uint8_t shifted_size_;
    storage_type storage_[N];

    size_t size() const { return shifted_size_ >> 1; }
  };

  union {
    // low bit will be set if small
    uint8_t discriminant_;
    DynamicState dynamic_;
    SmallState small_;
  };

  bool is_small() const { return discriminant_ & 1; }

  size_t size() const {
    return is_small() ? small_.size() >> 1 : dynamic_.size_;
  }
};

template <typename T, size_t N>
struct BigEndianState {
  using Storage = StaticVectorMixin<T>;
  using storage_type = typename Storage::storage_type;

  struct DynamicState {
    // high bit will be zero for discrimination
    uint32_t size_, capacity_;
    T* data_;
  };

  struct SmallState {
    // high bit will be one for discrimination
    uint8_t shifted_size_;
    storage_type storage_[N];

    size_t size() const { return shifted_size_ & 0b0111'1111; }
  };

  union {
    // high bit will be set if small
    uint8_t discriminant_;
    DynamicState dynamic_;
    SmallState small_;
  };

  bool is_small() const { return discriminant_ & 0b1000'0000; }

  size_t size() const {
    return is_small() ? small_.size() >> 1 : dynamic_.size_;
  }
};

template <typename T, size_t N>
struct SmallVector {
#ifdef BIGENDIAN
  using State = BigEndianState<T, N>;
#else
  using State = LittleEndianState<T, N>;
#endif
  State state_;

  using Storage = StaticVectorMixin<T>;
  using storage_type = typename Storage::storage_type;

  T* data() {
    return state_.is_small() ? Storage::ptr_at(state_.small_.storage_, 0)
                             : state_.dynamic_.data_;
  }

  ~SmallVector() {
    if (state_.is_small()) {
      Storage::destroy_storage(state_.small_.storage_, state_.small_.size());
    } else {
      delete[] state_.dynamic_.data_;
    }
  }

  // ...
};
```

Reply:
We can, but we must balance processing time against memory footprint as well. The main point of small vectors is to avoid dynamic allocations. But I don't think we'll have many of them in existence at any given point, so trying to save a few bytes doesn't sound worthwhile.

Reply:
I have the opposite intuition; I think it would frequently be handy to have potentially large numbers of them and stay efficient. For example, [the subtree pruning algorithm](https://github.com/bkietz/arrow/blob/31bc1825ac1dc8c4c02869f0c1dda10d3b20f805/cpp/src/arrow/compute/exec/subtree_internal.h#L58-L70)

Reply:
thousands to millions

Reply:
The way you have written it currently is the way LLVM's SmallVector works. In light of that I don't have the confidence to stand by my intuition for now.